packages/tre/src/plugins/cache/gateway.ts

```diff
@@ -130,137 +142,186 @@ function getMatchResponse(
     if (isSitesRequest(request)) throw new CacheMiss();
-
     cacheKey ??= request.url;
-    const cached = await this.storage.get<CacheMetadata>(cacheKey);
+
+    let resHeaders: Headers | undefined;
+    let resRanges: InclusiveRange[] | undefined;
+
+    const cached = await this.storage.get(cacheKey, ({ size, headers }) => {
+      resHeaders = new Headers(headers);
+      const contentType = resHeaders.get("Content-Type");
+
+      // Need size from metadata to parse `Range` header
+      const rangeHeader = request.headers.get("Range");
+      if (rangeHeader !== null) {
+        resRanges = _parseRanges(rangeHeader, size);
+        if (resRanges === undefined) throw new RangeNotSatisfiable(size);
+      }
+
+      return {
+        ranges: resRanges,
+        contentLength: size,
+        contentType: contentType ?? undefined,
+      };
+    });
     if (cached?.metadata === undefined) throw new CacheMiss();
 
-    const response = new CacheResponse(
-      cached.metadata,
-      cached.value
-    ).toResponse();
-    response.headers.set("CF-Cache-Status", "HIT");
+    // Avoid re-constructing headers if we already extracted multipart options
+    resHeaders ??= new Headers(cached.metadata.headers);
+    resHeaders.set("CF-Cache-Status", "HIT");
+    resRanges ??= [];
 
     return getMatchResponse(
       request.headers,
       cached.metadata.status,
-      response.headers,
-      cached.value
+      resHeaders,
+      resRanges,
+      cached.value,
+      cached.metadata.size
     );
   }
 
   async put(
     request: Request,
-    value: Uint8Array,
+    value: ReadableStream<Uint8Array>,
     cacheKey?: string
   ): Promise<Response> {
-    // Never cache Workers Sites requests, so we always return on-disk files
+    // Never cache Workers Sites requests, so we always return on-disk files.
     if (isSitesRequest(request)) return new Response(null, { status: 204 });
 
     const response = await HttpParser.get().parse(value);
+    let body = response.body;
+    assert(body !== null);
 
     const { storable, expiration, headers } = getExpiration(
       this.clock,
       request,
-      new Response(response.body, {
-        status: response.status,
-        headers: response.headers,
-      })
+      response
     );
-    if (!storable) {
-      throw new StorageFailure();
-    }
+    if (!storable) throw new StorageFailure();
 
     cacheKey ??= request.url;
-    await this.storage.put<CacheMetadata>(cacheKey, {
-      value: response.body,
-      expiration: millisToSeconds(this.clock() + expiration),
-      metadata: {
-        headers: Object.entries(headers),
-        status: response.status,
-      },
+
+    // If we know the size, avoid passing the body through a transform stream to
+    // count it (trusting `workerd` to send correct value here).
+    // Safety of `!`: `parseInt(null)` is `NaN`
+    const contentLength = parseInt(response.headers.get("Content-Length")!);
+    let sizePromise: Promise<number>;
```

Comment:
Completely stylistic, but I'm not a huge fan of variable names including the type